Smart Select language service API

andrewbranch opened the pull request:
<!--
Thank you for submitting a pull request!

Here's a checklist you might find useful.
* [ ] There is an associated issue that is labeled 'Bug' or 'help wanted'
* [ ] Code is up-to-date with the `master` branch
* [ ] You've successfully run `gulp runtests` locally
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Closes #29071 

### Protocol for VS & VS Code

```ts
export interface Location {
    line: number;
    offset: number;
}

export interface TextSpan {
    start: Location;
    end: Location;
}

export interface SelectionRangeRequest extends FileRequest {
    command: CommandTypes.SelectionRange; // 'selectionRange'
    arguments: SelectionRangeRequestArgs;
}

export interface SelectionRangeRequestArgs extends FileRequestArgs {
    locations: Location[];
}

export interface SelectionRangeResponse extends Response {
    body?: SelectionRange[];
}

export interface SelectionRange {
    textSpan: TextSpan;
    parent?: SelectionRange;
}
```

Also exposes a `selectionRange-full` tsserver command which accepts the same request format but responds with the language service response unchanged, which is the same except that `TextSpan` is a `ts.TextSpan`:

```ts
export interface TextSpan {
    start: number;
    length: number;
}
```

### Implementation notes

Selection ranges walk up the AST, with a few different types of exceptions to the rule.

#### Adding artificial depth to flat nodes
Some TypeScript nodes _feel_ like they have subtrees but are actually quite flat. Keeping them flat results in a jump from one selection to the next that feels too big. To remedy this, when asking for the children of these nodes, we synthesize SyntaxLists as intermediate nodes that group children together more intuitively, sometimes multiple levels deep. The most complex example is for mapped types. Consider

```ts
type M = { -readonly [K in keyof any]-?: any }
```

The actual tree for the type in this declaration looks like

```
MappedType
  ├ OpenBraceToken
  ├ MinusToken
  ├ ReadonlyToken
  ├ OpenBracketToken
  ├ TypeParameter
  │  ├ Identifier
  │  ├ InKeyword
  │  └ TypeOperator
  │      ├ KeyofKeyword
  │      └ AnyKeyword
  ├ CloseBracketToken
  ├ MinusToken
  ├ QuestionToken
  ├ ColonToken
  ├ AnyKeyword
  └ CloseBraceToken
```

But this belies the intuition of the way things are grouped. The two tokens in `-readonly` go together, the two in `-?` go together, the square brackets go with their contents, etc. The tree we actually want for selection purposes would look more like

```
MappedType
  ├ OpenBraceToken
  ├ SyntaxList
  │  ├ SyntaxList
  │  │  ├ MinusToken
  │  │  └ ReadonlyToken
  │  ├ SyntaxList
  │  │  ├ OpenBracketToken
  │  │  ├ TypeParameter
  │  │  │  ├ Identifier
  │  │  │  ├ InKeyword
  │  │  │  └ TypeOperator
  │  │  │     ├ KeyofKeyword
  │  │  │     └ AnyKeyword
  │  │  └ CloseBracketToken
  │  ├ SyntaxList
  │  │  ├ QuestionToken
  │  │  └ ColonToken
  │  └ AnyKeyword
  └ CloseBraceToken
```

This results in a better selection experience, but it's totally fake which also makes it totally subjective. I used two main strategies to make decisions that are ultimately opinions:

1. As someone who writes a lot of TypeScript in VS Code myself, I put myself in the shoes of a user selecting some of my own code and thought about the common edits I might want to make around a particular selection, and tried to add selection stops that facilitated those edits with minimal additional keystrokes.
2. I gauged some of my own opinions and expectations that weren’t reflected in the AST against @mjbvz’s. Where we agreed, I felt confident that synthesizing a node was justified.

#### Skipping redundant or uninteresting nodes
Identical selection ranges are deduped, but some others vary only in whitespace or semicolons and seemed to add little value. These nodes are skipped.

#### Adding JSDoc comments
An extra selection is added to nodes with JSDoc annotations so that the node can be selected both alone and with its comment.

#### Special handling for string literals
String literals don't separate their quotes from their contents in the AST, which is desirable for selection, so there’s special logic there. Template strings are especially unintuitive and have special handling to make sure you can get the whole string, just inside the quotes, around the outside of a substitution (`${x}`), and on the inside of a substitution.

amcasey commented on src/server/protocol.ts:
```diff
@@ -130,7 +130,10 @@ namespace ts.server.protocol {
+        ConfigurePlugin = "configurePlugin",
+        SelectionRange = "selectionRange",
+        /* @internal */
+        SelectionRangeFull = "selectionRange-full",
```
Thank you :smile:

mjbvz commented:
Protocol changes look good. I'll give it another quick test locally and let you know if I run into an big issues

mjbvz commented:
Looks great. Just found a few odd results that cause VS Code to error:

---
```js
export function loadRepos() {
  return {
    type: /**/LOAD_REPOS,
  };
}
```

Top level returned text span is:

```
        "textSpan": {
            "start": {
                "line": 1,
                "offset": 27
            },
            "end": {
                "line": 1,
                "offset": 27
            }
        },
```

---

```js
describe('thing', () => {
    it('/**/', () => {});
});
```

Returned span is:

```
        "textSpan": {
            "start": {
                "line": 1,
                "offset": 20
            },
            "end": {
                "line": 1,
                "offset": 20
            }
        },
```

---

```js
class HomePage {
    componentDidMount() {
        if (this.props.username/**/) { }
    }
}
```

Returned top level span is

```
 "textSpan": {
            "start": {
                "line": 2,
                "offset": 23
            },
            "end": {
                "line": 2,
                "offset": 23
            }
        },
```

---

andrewbranch commented:
Hm, I can see why the empty string one is doing that, but the others seem odd. I'll check and see what it thinks it's doing, but it looks like I just need to add a filter to reject zero-width spans. Will update first thing tomorrow. Thanks for testing!

andrewbranch commented:
Ohhh. Those really strange ones were caused by a possible bug in `ts.positionBelongsToNode` that @RyanCavanaugh and @rbuckton and I were discussing yesterday. I've since switched away from using that API, and just added a filter to ignore empty spans.

I can no longer repro the bugs you gave me, but I thought I stopped using `ts.positionBelongsToNode` before you tried it yesterday... I just added unit tests for those cases but I'll try the integration with your custom VS Code build too.

sheetalkamat commented on src/testRunner/unittests/tsserver/smartSelection.ts:
```diff
@@ -0,0 +1,687 @@
+
+    describe("unittests:: tsserver:: smartSelection", () => {
+        it("works for simple JavaScript", () => {
+            const getSmartSelectionRange = setup("/file.js", `
```
So having couple of tests here to verify tsserver API is good idea but rest of them should be fourslash test as its mainly language service functionality you are testing.

andrewbranch replied:
What exactly is the motivation for moving them to fourslash? Having them here, with the code in template strings, has the advantage that you can read line and offset from your editor. In fourslash, offset would be difficult to use because of the four slashes and marker characters. As an alternative, I wrote a utility to convert these existing tests to fourslash ranges, thinking that might be easier to write and read, but I think it's worse, at least to read:

```
[|type M = [|{ [|[|-readonly [K in keyof any][|-[|/**/?|]|]|]: any|] }|];|]|]
```

I get that fourslash is one step closer to the language service, but I'm not sure it's a good format for assertions about this. Have any suggestions?

andrewbranch replied:
@sheetalkamat done. Created a baseline format for these.

RyanCavanaugh commented:
I agree the fourslash format looks like a disaster.

The motivation, though, is that in general position-based tests end up being a lot less readable and maintainable than DSL-based formats, so we try to avoid the former as much as possible.

It's impossible to look at these tests right now in the diff view, for example, and understand what the expected behavior is, and if something changes in the wire format, there's a bunch of updating that will have to happen. Updating the text is also very difficult - adding a single character in the middle of an existing test is going to require a lot of tedious math.

Having a one-off DSL has generally paid for itself in terms of long-term test maintenance. For example, we could imagine having something like this (bikeshed a bunch):
```ts
const test = `const { x, y: a, ...zs = {} } = {};`;
const expected = [
    `const { x, y: a, ...zs = {} } = {};`,
    `             [a]                    `,
    `          [y: a]                   `,
    `     [{ x, y: a, ...zs = {} }]     `,
];
```

Alternatively, this might be an appropriate place to use baselines that just render a human-readable format of what happened. Writing a one-off renderer to make an inspectable text file for what the selection spans were would be pretty easy, and lends itself to reviewers being able to see what the behavior is.

andrewbranch commented:
👍🏽 I like the idea of a baseline. The assertions are difficult to _write_ no matter what, but there are formats that are easy to _read_—eliminating the writing step seems like the best approach.

DanielRosenwasser commented on src/services/smartSelection.ts:
```diff
@@ -0,0 +1,266 @@
+     * @param sourceFile The source file containing the nodes.
+     */
+    function positionShouldSnapToNode(pos: number, node: Node, nextNode: Node | undefined) {
+        // Can’t use 'ts.positionBelongsToNode()' here because it cleverly accounts
```
Why do these files have smart quotes?

andrewbranch replied:
I habitually [type proper apostrophes by hand](https://twitter.com/atcb/status/1119075745446256640) when writing English sentences, including in code comments. (OS-level quote replacement tends to be unreliable, and hard to fight when it’s wrong, e.g. figures like `50° 32' 24.1"` or `6' 2"` require straight hatches. Bad typography drives me bananas so at some point I disabled all the auto-replacement features and started typing the character I mean.)

Is there some environment where this causes a problem? I’ve never had any issues, nor has anyone ever noticed or pointed it out 😄

DanielRosenwasser replied:
There's probably nothing wrong with it, but it does feel unusual. I've had issues where one team hands off an HTML file with content that was copy/pasted from Word and that caused issues for one reason or another. So I guess I'm scarred from that.

andrewbranch replied:
Don’t worry; you can count on me never to paste content from Word into the TypeScript repo 😛

andrewbranch commented:
Ryan just gave me a physical thumbs up on this in the room because his computer currently has a blue screen of death 😄

sheetalkamat commented:
@andrewbranch yey on first big feature merge 👍

mihailik commented:
External observer question: why is `selectionRange-full` is needed?

>```
>    export interface TextSpan {
>       start: Location;
>       end: Location;
>   }
>```

. . .

> Also exposes a selectionRange-full tsserver command which accepts the same request format but responds with the language service response unchanged, which is the same except that TextSpan is a ts.TextSpan:
>
>```
>  export interface TextSpan {
>      start: number;
>      length: number;
>  }
>```

Given general brevity of the docs on ts APIs, adding odd-shaped nearly-identical calls can easily lead people astray.

andrewbranch commented:
Good question. Almost every TSServer command has a `-full` counterpart that gets used by Visual Studio. I believe Visual Studio internally uses 0-based lines and columns, while I _know_ VS Code uses 1-based lines and columns.

I actually wrote up something about this for the wiki, I just haven’t added it yet.

mihailik commented:
Sorry, I didn't realise it's already consistent with other calls. Makes total sense!

amcasey commented:
I believe the primary difference between the `-full` and regular versions is that `-full` uses a character offset from the beginning of the file, rather than line and column.

andrewbranch commented:
Yep, that’s right, and that’s the same way that the compiler itself typically represents positions in files. Maybe Visual Studio doesn't use lines and columns much internally at all ¯\_(ツ)_/¯
